module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,104 @@
+#include "advancedCalculator.hpp"
+
+#include <cmath>
+#include <functional>
+#include <map>
+#include <regex>
+
+#include <iostream>
+
+const std::map<char, std::function<double(double, double)>> operations{
+    {'+', std::plus<>()},
+    {'-', std::minus<>()},
+    {'*', std::multiplies<>()},
+    {'/', std::divides<>()},
+    {'%', std::modulus<int>()},
+    {'^', [](auto base, auto exponent) { return pow(base, exponent); }},
+    {'$', [](auto value, auto root) { return pow(value, 1.0 / root); }},
+    {'!', [](auto value, auto notUsed) { return value <= 0 ? -std::tgamma(1.0 - value) : std::tgamma(value + 1.0); }}};
+
+const std::regex regex2arguments{R"((^[\-]?\d+\.*\d*)\s*(\D)\s*([\-]?\d+\.*\d*)$)"};
+const std::regex regex1argument{R"((^[\-]?\d+\.*\d*)\s*(\D)$)"};
+
+static bool parseString(const std::string& string, char* operation, double* lhs, double* rhs)
+{
+    if (std::smatch match;
+        std::regex_search(string, match, regex2arguments) or std::regex_search(string, match, regex1argument)) {
+        if (match[0] != string) {
+            return false;
+        } else {
+            *lhs = stod(match[1]);
+            *operation = ((std::string)match[2])[0];
+            if (match.size() == 4) {
+                *rhs = stod(match[3]);
+            }
+            if ((match.size() == 4) and (*operation == '!')) {
+                return false;
+            }
+            return true;
```

Comment:
```suggestion
            return (match.size() == 4) and (*operation == '!')
```
Small simplification.